node/modules/storageminer.go

```diff
@@ -357,78 +357,193 @@ func SectorStorage(mctx helpers.MetricsCtx, lc fx.Lifecycle, lstor *paths.Local,
+			p.backOff(ctx)
+		case progress.Instance+2 >= lease.ValidityTerm:
+			log.Infof("F3 progressed (%d) to within two instances of lease expiry (%d). Restarting participation.", progress.Instance, lease.ValidityTerm)
+			return nil
+		default:
+			remainingInstanceLease := lease.ValidityTerm - progress.Instance
+			log.Debugf("F3 participation lease is valid for further %d instances. Recking after %s.", remainingInstanceLease, p.checkProgressInterval)
+			p.backOffFor(ctx, p.checkProgressInterval)
+		}
+	}
+	return ctx.Err()
+}
 
-			// we have succeeded in giving a lease, reset the backoff
-			b.Reset()
+func (p *f3Participator) backOff(ctx context.Context) {
+	p.backOffFor(ctx, p.backoff.Duration())
+}
 
-			oldLease = newLease
-			// wait for the half of the lease time and then refresh
-			timer.Reset(leaseTime / 2)
-			select {
-			case <-ctx.Done():
-				return
-			case <-timer.C:
-			}
+func (p *f3Participator) backOffFor(ctx context.Context, d time.Duration) {
+	// Create a timer every time to avoid potential risk of deadlock or the need for
+	// mutex despite the fact that f3Participator is never (and should never) be
+	// called from multiple goroutines.
+	timer := time.NewTimer(d)
+	select {
+	case <-ctx.Done():
+		return
+	case <-timer.C:
+	}
+}
+
+func F3Participation(mctx helpers.MetricsCtx, lc fx.Lifecycle, node v1api.FullNode, participant dtypes.MinerAddress) error {
+	const (
+		// maxCheckProgressAttempts defines the maximum number of failed attempts
+		// before we abandon the current lease and restart the participation process.
+		//
+		// The default backoff takes 12 attempts to reach a maximum delay of 1 minute.
+		// Allowing for 13 failures results in approximately 2 minutes of backoff since
+		// the lease was granted. Given a lease validity of up to 5 instances, this means
+		// we would give up on checking the lease during its mid-validity period;
+		// typically when we would try to renew the participation ticket. Hence, the value
+		// to 13.
+		checkProgressMaxAttempts = 13
+		// checkProgressInterval defines the duration between progress checks in normal operation mode.
+		// This interval is used when there are no errors in retrieving the current progress.
+		checkProgressInterval = 10 * time.Second
+		// leaseTerm The number of instances the miner will attempt to lease from nodes.
+		leaseTerm = 5
+	)
+	ctx, cancel := context.WithCancel(mctx)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		participator := newF3Participator(
+			node,
+			participant,
+			&backoff.Backoff{
+				Min:    1 * time.Second,
+				Max:    1 * time.Minute,
+				Factor: 1.5,
+			},
+			checkProgressMaxAttempts,
+			checkProgressInterval,
+			leaseTerm,
+		)
+
+		switch err := participator.participate(ctx); {
+		case err == nil, errors.Is(err, context.Canceled):
+			log.Infof("Stopped participating in F3")
+		default:
+			log.Errorw("F3 participation stopped abruptly", "err", err)
 		}
```

Comment:
can we wait for `OnStart` here? It would also be kind of nice to write this more like a service (with start/stop methods).

Reply:
@Stebalien could you check my [handwriting](https://github.com/filecoin-project/lotus/pull/12531/commits/ec3b3f03950659872dbb69acd5fbcf19bf091a9f) on this one please?